doc/developer/design/20220516_transactional_consistency.md

```diff
@@ -0,0 +1,167 @@
+# Materialize Transactional Consistency
+
+## Summary
+
+A description of Materialize’s consistency guarantees with respect to the operations that occur inside of Materialize.
+Operations include: `SELECT`, `INSERT`, `UPDATE`, and `DELETE` statements (but not `TAIL`)
+
+This design document is in service of [Epic #11631](https://github.com/MaterializeInc/materialize/issues/11631)
+and [Correctness Property #1](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/platform/ux.md#correctness)
+. Specifically for the first bullet point from the correctness property, “`SELECT`, `INSERT`, `UPDATE`, and `DELETE`
+statements (but not `TAIL`)". A follow-up design document will be created for the second bullet point, "Materialize
+initiated acknowledgements of upstream sources (e.g., the commit of an offset by a Kafka source, the acknowledge of an
+LSN by a PostgresSQL source, etcetera)".
+
+This document does not take any opinion on whether the changes proposed should be the default behavior or an opt-in
+behavior via some configuration.
+
+## Goals
+
+The goal of this design document is to describe what changes are needed to provide Strict Serializability to the
+operations mentioned above.
+
+## Non-Goals
+
+Sinks are left out of this document and not considered for consistency.
+
+`TAIL` is left out of this document and not considered for consistency.
+
+`AS OF` allows clients to specify the timestamp of a query. Queries with `AS OF` do not constrain the timestamp
+selection of other queries. These properties allow clients to circumvent our consistency guarantees, so `AS OF` is left
+out of this document.
+
+This document does not include Materialize initiated acknowledgements of upstream sources (e.g., the commit of an offset
+by a Kafka source, the acknowledgment of an LSN by a PostgresSQL source, etcetera). For the purposes of this document we
+will consider all of these events to have happened at the timestamp assigned to them by timely workers. This is not
```

Comment:
The previous discussion is here: https://github.com/MaterializeInc/materialize/pull/12475#discussion_r875015796

I'm trying to highlight the following scenario:
1. We receive a transaction, `txn`, from Postgres, assign it a timestamp of `ts + 100`, ingest it, and then send an ACK back to Postgres.
2. We read some view over the Postgres source at timestamp `ts` and don't see the results of `txn`.

This is a violation of strict serializability because the read happens after the ingestion in real time, but the read doesn't see `txn`. The design doc as it's written does not prevent this scenario, so I'm just trying to point out that we're ok with this for now.

The connection for me is that I view receiving data from an upstream source and sending an acknowledgement for that data to all be part of a single operation as it relates to consistency, where the timestamp assigned to that data is used to order that operation. Essentially, ingesting and ACKing data from an upstream source are all part of a single write transaction where the client is the upstream source.

My motivating example is Postgres (or any other DB) synchronous replication: 
1. A client sends a write to PG.
2. PG sends the write to MZ.
3. MZ receives the write and assigns it a timestamp.
4. MZ ingests the txn.
5. MZ sends ACK to PG.
6. PG sends ACK to client.

This would help ensure to users that if their transaction on an upstream source commits, then the data will be readable in MZ.

I may be misunderstanding this and I could also probably make this clearer in the doc.